src/types/utils.cpp

```diff
@@ -158,5 +166,105 @@ symbolic::Expression get_type_size(const types::IType& type, bool allow_comp_tim
     }
 }
 
+std::unique_ptr<typename types::IType> infer_type_from_container(
+    analysis::AnalysisManager& analysis_manager, const StructuredSDFG& sdfg, std::string container
+) {
+    if (sdfg.type(container).type_id() == types::TypeID::Scalar) {
+        return sdfg.type(container).clone();
+    }
+
+    std::unique_ptr<typename types::IType> type = nullptr;
+    auto& users = analysis_manager.get<analysis::Users>();
+
+    for (auto user : users.reads(container)) {
+        if (auto access_node = dynamic_cast<data_flow::AccessNode*>(user->element())) {
+            for (auto& memlet : user->parent()->out_edges(*access_node)) {
+                if (type == nullptr) {
```

Comment:
Since you can focus on pointers only, you can simplify the body to

```cpp
auto& base_type = memlet.base_type();
if (type == nullptr) {
   type = &base_type;
   continue;
}

if (*type != base_type) {
    return nullptr;
}
```